File: backend/app/api/routes/analyses.py
"""Analysis routes — POST and GET /api/v1/analyses."""

from fastapi import APIRouter, Query

from app.core.logging import get_logger
from app.schemas.analysis import (
    AnalysisListResponse,
    AnalysisResponse,
    CreateAnalysisRequest,
)
from app.services.analysis_service import analysis_service
from app.repositories.analysis_repository import analysis_repository

logger = get_logger("routes.analyses")

router = APIRouter(tags=["Analyses"])


@router.post("/analyses", response_model=AnalysisResponse, status_code=201)
async def create_analysis(body: CreateAnalysisRequest) -> AnalysisResponse:
    """Run a full plant disease analysis for a previously uploaded image.

    Orchestrates:
    1. Image metadata retrieval
    2. Product quality check (Sprint A: deterministic stub)
    3. Mock hierarchical inference pipeline (Sprint A: deterministic adapter)
    4. MongoDB persistence
    5. Structured response
    """
    doc = analysis_service.create_analysis(image_id=body.image_id)
    return AnalysisResponse(**doc)


@router.get("/analyses/{analysis_id}", response_model=AnalysisResponse)
async def get_analysis(analysis_id: str) -> AnalysisResponse:
    """Retrieve a single analysis by its UUID identifier."""
    doc = analysis_repository.get_by_id(analysis_id)
    return AnalysisResponse(
        analysis_id=doc["id"],
        status=doc["status"],
        image=doc["image"],
        quality=doc.get("quality"),
        leaf_check=doc.get("leaf_check"),
        crop_check=doc.get("crop_check"),
        prediction=doc.get("prediction"),
        created_at=doc["created_at"],
        updated_at=doc["updated_at"],
    )


@router.get("/analyses", response_model=AnalysisListResponse)
async def list_analyses(
    skip: int = Query(default=0, ge=0, description="Number of records to skip"),
    limit: int = Query(default=20, ge=1, le=100, description="Maximum records to return"),
) -> AnalysisListResponse:
    """Return a paginated list of analyses ordered newest first."""
    items, total = analysis_repository.list(skip=skip, limit=limit)

    response_items = [
        AnalysisResponse(
            analysis_id=doc["id"],
            status=doc["status"],
            image=doc["image"],
            quality=doc.get("quality"),
            leaf_check=doc.get("leaf_check"),
            crop_check=doc.get("crop_check"),
            prediction=doc.get("prediction"),
            created_at=doc["created_at"],
            updated_at=doc["updated_at"],
        )
        for doc in items
    ]

    return AnalysisListResponse(
        items=response_items,
        total=total,
        skip=skip,
        limit=limit,
    )

File: frontend/lib/api/analyses.js
import { apiRequest } from "./client";

/**
 * Submit an analysis request for a validated image.
 *
 * @param {string} imageId - UUID string of uploaded image
 * @returns {Promise<Object>} AnalysisResponse object
 */
export async function createAnalysis(imageId) {
  return apiRequest("/analyses", {
    method: "POST",
    body: { image_id: imageId },
  });
}

/**
 * Retrieve a stored analysis record by ID.
 *
 * @param {string} analysisId - UUID string of analysis
 * @returns {Promise<Object>} AnalysisResponse object
 */
export async function getAnalysis(analysisId) {
  return apiRequest(`/analyses/${encodeURIComponent(analysisId)}`);
}

/**
 * List paginated analysis records for current user.
 *
 * @param {Object} params - { skip, limit }
 * @returns {Promise<Object>} AnalysisListResponse { items, total, skip, limit }
 */
export async function listAnalyses({ skip = 0, limit = 20 } = {}) {
  const query = new URLSearchParams({
    skip: String(skip),
    limit: String(limit),
  });
  return apiRequest(`/analyses?${query.toString()}`);
}
